File: cressent/modules/tanglegram.R
#!/usr/bin/env Rscript

# Load necessary libraries
suppressPackageStartupMessages({
  library(ape)       # For tree reading and manipulation
  library(dendextend) # For tanglegrams
  library(DECIPHER)
})


# Define CLI arguments
# Parse command-line arguments
args <- commandArgs(trailingOnly = TRUE)
# Define default values for arguments
arg_list <- list(
    tree1 = NULL,
    tree2 = NULL,
    label1 = "Tree 1",
    label2 = "Tree 2",
    output = "./",
    name_tanglegram = "tanglegram.pdf",
    height = 11,
    width = 20,
    lab.cex=1.5
)

# Assign values from command-line arguments
for (arg in args) {
  key_value <- strsplit(arg, "=")[[1]]
  if (length(key_value) == 2) {
    arg_list[[key_value[1]]] <- key_value[2]
  }
}

# Check for required arguments
if (is.null(arg_list$tree1) || is.null(arg_list$tree2)) {
  stop("Error: --tree1 and --tree2 are required arguments.")
}


# Assign arguments to variables
path1 <- arg_list$tree1
path2 <- arg_list$tree2
label1 <- arg_list$label1
label2 <- arg_list$label2
output_dir <- arg_list$output
output_name <- arg_list$name_tanglegram
height <- as.numeric(arg_list$height)
width <- as.numeric(arg_list$width)
lab.cex <- as.numeric(arg_list$lab.cex)

# Load the trees
dend1 <- ReadDendrogram(path1)
dend2 <- ReadDendrogram(path2)

# Extract labels from dendrogram on the left
labels1 <- dend1 %>% set("labels_to_char") %>% labels 
labels2 <- dend2 %>% set("labels_to_char") %>% labels 

# Check for common labels and prune trees if needed
common_labels <- intersect(labels1, labels2)
if (length(common_labels) == 0) {
  stop("Error: The two trees have no common labels.")
} 

# Create output directory if it does not exist
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# Create tanglegram
# Calculate the maximum number of characters in the tree labels
max_label_length <- max(
  max(nchar(labels1)), 
  max(nchar(labels2))
)
# Dynamically set margin_inner based on label length
margin_inner <- max(max_label_length * .5)
# Adjust the Margins
num_labels <- max(length(labels1), length(labels2))

# createa dendrogram list
dl <- dendlist(dend1, dend2)

# # calculate RF score (Robinson-Foulds distance)
RF_score = dist.dendlist(dl, method = c("edgeset"))
print(RF_score)
# save the plot
# Dynamically Adjust the PDF Dimensions
pdf(file = file.path(output_dir, output_name), width = max(width, num_labels / 10), height = max(height, num_labels / 15))
# pdf(file = file.path(output_dir, output_name), width = width, height = height)
dl %>% 
    # dendextend::untangle(method = "step2side") %>%
    tanglegram(main_left = label1,
            main_right = label2,
            main = paste("RF score:", RF_score),
            highlight_distinct_edges = TRUE,
            highlight_branches_col = TRUE,
            common_subtrees_color_branches = TRUE,
            highlight_branches_lwd=TRUE,
            # margin_inner = margin_inner,
            lwd = 2,
            lab.cex=lab.cex)
# dev.copy(pdf, file.path(output_dir, output_name))
dev.off()


cat(sprintf("Tanglegram saved to %s\n", file.path(output_dir, output_name)))

File: cressent/modules/tanglegram.py
#!/usr/bin/env python3

import argparse
import subprocess
import os
import re
from pathlib import Path


def run_command(command, error_message):
    """
    Helper function to run a shell command and handle errors.
    """
    try:
        subprocess.run(command, shell=False, check=True)
        print(f"Command succeeded: {command}")
    except subprocess.CalledProcessError as e:
        print(f"{error_message}: {e}")
        exit(1)

def generate_tanglegram(tree1, tree2, label1, label2, output_dir, name_tanglegram="tanglegram.pdf", width=20, height=11,lab_cex=1.5):
    """
    Generate a tanglegram using an R script.
    
    Args:
        tree1 (str): Path to the first tree file.
        tree2 (str): Path to the second tree file.
        label1 (str): Label for the first tree.
        label2 (str): Label for the second tree.
        output_dir (str): Directory to save the output.
        name_tanglegram (str): Name of the output PDF file (default: "tanglegram.pdf").
        width (numeric): width of the pdf (default: 20 units).
        height (numeric): height of the pdf (default: 11 units).
        lab_cex (numeric): cex size of the labels.
    """
    # Path to the R script
    modules_dir = Path(__file__).resolve().parent  # Directory where the script is located
    r_script_path = modules_dir / "tanglegram.R"  # Update with the actual path
    #r_script_path = "/fs/project/PAS1117/ricardo/ssDNA_tool/ssDNA_annotator/modules/tanglegram.R"  # Update with the actual path

    # Build the command
    cmd = [
        "Rscript",
        r_script_path,
        f"tree1={tree1}",
        f"tree2={tree2}",
        f"label1={label1}",
        f"label2={label2}",
        f"output={output_dir}",
        f"name_tanglegram={name_tanglegram}",
        f"height={height}",
        f"width={width}"
        f"lab.cex={lab_cex}"
    ]

    # Run the R script
    run_command(cmd, "Error making tanglegram")

def main():
    parser = argparse.ArgumentParser(description = "Generate a tanglegram from two phylogenetic trees.")
    parser.add_argument("--tree1", required = True, help = "Path to the first tree file.") # arg_tree1
    parser.add_argument("--tree2", required = True, help = "Path to the second tree file.") # arg_tree2
    parser.add_argument("--label1", required = True, help = "Label for the first tree in the tanglegram.") # arg_label1
    parser.add_argument("--label2", required = True, help = "Label for the second tree in the tanglegram.") # arg_label2
    parser.add_argument("-o", "--output", required=True, default=".",help="Path to the output directory (Default: working directory)") # arg_output
    parser.add_argument("--name_tanglegram", default = "tanglegram.pdf", help = "Name of the tanglegram (default: tanglegram.pdf)") 
    parser.add_argument("--width", default = 20, type=float, help = "Width of the tanglegram (default = 20)") 
    parser.add_argument("--height", default = 11, type=float, help = "Height of the tanglegram (default = 11)") 
    parser.add_argument("--lab_cex", default = 1.5, type=float, help = "cex size of the labels (default = 1.5)") 

    args = parser.parse_args()

    output_dir = args.output
    os.makedirs(output_dir, exist_ok=True)
    
    # run R script
    generate_tanglegram(args.tree1, args.tree2, args.label1, args.label2, output_dir, name_tanglegram=args.name_tanglegram, width=args.width, height=args.height, lab_cex=args.lab_cex)


if __name__ == "__main__":
    main()


